File: solutions/java/code-carl/ch08_greedy/T2_LC0376_WiggleMaxLength.java
/**
 * Practice template for:
 * - LeetCode 376. Wiggle Subsequence
 *   https://leetcode.cn/problems/wiggle-subsequence/
 */
public class T2_LC0376_WiggleMaxLength {
    public int wiggleMaxLength(int[] nums) {
        int preDiff = 0;
        int currDiff;
        int count = 1;

        for (int i = 0; i < nums.length - 1; i++) {
            currDiff = nums[i + 1] - nums[i];
            if (preDiff >= 0 && currDiff < 0 || preDiff <= 0 && currDiff > 0) {
                count++;
                preDiff = currDiff;
            }
        }
        return count;
    }
}

File: templates/python/code-carl/ch08_greedy/t2_lc0376_wiggle_max_length.py
"""LeetCode 376. Wiggle Subsequence
https://leetcode.cn/problems/wiggle-subsequence/
"""

from typing import List


class Solution:
    def wiggleMaxLength(self, nums: List[int]) -> int:
        pass

if __name__ == "__main__":
    solution = Solution()
    assert solution.wiggleMaxLength([1, 7, 4, 9, 2, 5]) == 6
    assert solution.wiggleMaxLength([1, 17, 5, 10, 13, 15, 10, 5, 16, 8]) == 7
    assert solution.wiggleMaxLength([1, 2, 3, 4, 5, 6, 7, 8, 9]) == 2
    print("All tests passed.")
